project_clusters.go

```diff
@@ -0,0 +1,204 @@
+	CreatedAt          *time.Time         `json:"created_at,omitempty"`
+	ProviderType       string             `json:"provider_type,omitempty"`
+	PlatformType       string             `json:"platform_type,omitempty"`
+	EnvironmentScope   string             `json:"environment_scope,omitempty"`
+	ClusterType        string             `json:"cluster_type,omitempty"`
+	User               *User              `json:"user,omitempty"`
+	PlatformKubernetes PlatformKubernetes `json:"platform_kubernetes,omitempty"`
+	Project            *Project           `json:"project,omitempty"`
+}
+
+func (v ProjectCluster) String() string {
+	return Stringify(v)
+}
+
+// PlatformKubernetes represents a GitLab Project Cluster PlatformKubernetes.
+type PlatformKubernetes struct {
+	APIURL            string `json:"api_url,omitempty"`
+	Token             string `json:"token,omitempty"`
+	CaCert            string `json:"ca_cert,omitempty"`
+	Namespace         string `json:"namespace,omitempty"`
+	AuthorizationType string `json:"authorization_type,omitempty"`
+}
+
+// ListClusters gets a list of all clusters in a project.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/project_clusters.html#list-project-clusters
+func (s *ProjectClustersService) ListClusters(pid interface{}, options ...OptionFunc) ([]*ProjectCluster, *Response, error) {
+	projectID, err := parseID(pid)
+	if err != nil {
+		return nil, nil, err
+	}
+	u := fmt.Sprintf("projects/%s/clusters", url.QueryEscape(projectID))
+
+	req, err := s.client.NewRequest("GET", u, nil, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var vs []*ProjectCluster
+	resp, err := s.client.Do(req, &vs)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return vs, resp, err
+}
+
+// GetCluster gets a cluster.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/project_clusters.html#get-a-single-project-cluster
+func (s *ProjectClustersService) GetCluster(pid interface{}, clusterID int, options ...OptionFunc) (*ProjectCluster, *Response, error) {
+	projectID, err := parseID(pid)
+	if err != nil {
+		return nil, nil, err
+	}
+	u := fmt.Sprintf("projects/%s/clusters/%d", url.QueryEscape(projectID), clusterID)
+
+	req, err := s.client.NewRequest("GET", u, nil, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var vs *ProjectCluster
+	resp, err := s.client.Do(req, &vs)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return vs, resp, err
+}
+
+// AddClusterOptions represents the available AddCluster() options.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/project_clusters.html#add-existing-cluster-to-project
+type AddClusterOptions struct {
+	Name               string             `json:"name,omitempty"`
+	Enabled            *bool              `json:"enabled,omitempty"`
+	EnvironmentScope   string             `json:"environment_scope,omitempty"`
+	PlatformKubernetes PlatformKubernetes `json:"platform_kubernetes_attributes,omitempty"`
```

Comment:
Why would I add `url` tags here, since GitLab accepts this as json and not as query string?

Reply:
Because some options struct are also used in `GET` calls, in which case these fields have to be added as query params. So to keep things consistent and to prevent the need to check how every individual option struct can be used, we decided to always set these tags for all options structs.